controllers/service.go

```diff
@@ -0,0 +1,123 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Casdoor will expose its providers as services to SDK
+// We are going to implement those services as APIs here
+
+package controllers
+
+import (
+	"encoding/json"
+
+	"github.com/casbin/casdoor/object"
+	"github.com/casbin/casdoor/util"
+	sender "github.com/casdoor/go-sms-sender"
+)
+
+// SendEmail
+// This API is not for Casdoor frontend to call,
+// it is for Casdoor SDKs.
+func (c *ApiController) SendEmail() {
+	app := c.GetApplicationByClientIdAndSecret()
+	if app == nil {
+		c.ResponseError("Invalid clientId or clientSecret.")
+		return
+	}
+
+	provider := app.GetEmailProvider()
+	if provider == nil {
+		c.ResponseError("No Email provider for this application.")
+		return
+	}
+
+	var emailForm struct {
+		Title     string   `json:"title"`
+		Content   string   `json:"content"`
+		Receivers []string `json:"receivers"`
+		Sender    string   `json:"sender"`
+	}
+
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &emailForm)
+	if err != nil {
+		c.ResponseError("Request body error.")
+		return
+	}
+
+	if util.IsStrsEmpty(emailForm.Title, emailForm.Content, emailForm.Sender) {
+		c.ResponseError("Missing parameters.")
+		return
+	}
+
+	ok := 0
+	for _, receiver := range emailForm.Receivers {
+		if msg := object.SendEmail(
+			provider,
+			emailForm.Title,
+			emailForm.Content,
+			receiver,
+			emailForm.Sender);
+			len(msg) == 0 {
+			ok++
+		}
+	}
+
+	c.Data["json"] = Response{Status: "ok", Data: ok}
+	c.ServeJSON()
+}
+
+// SendSMS
+// This API is not for Casdoor frontend to call,
+// it is for Casdoor SDKs.
+func (c *ApiController) SendSMS() {
+	app := c.GetApplicationByClientIdAndSecret()
+	if app == nil {
+		c.ResponseError("Invalid clientId or clientSecret.")
+		return
+	}
+
+	provider := app.GetSmsProvider()
+	if provider == nil {
+		c.ResponseError("No Email provider for this application.")
+		return
+	}
+
+	client := sender.NewSmsClient(
+		provider.Type,
+		provider.ClientId,
+		provider.ClientSecret,
+		provider.SignName,
+		provider.RegionId,
+		provider.TemplateCode,
+		provider.AppId,
+	)
+	if client == nil {
+		c.ResponseError("Invalid provider info.")
+		return
+	}
+
+	var smsForm struct {
+		Receivers  []string          `json:"receivers"`
+		Parameters map[string]string `json:"parameters"`
+	}
+
+	err := json.Unmarshal(c.Ctx.Input.RequestBody, &smsForm)
+	if err != nil {
+		c.ResponseError("Request body error.")
+		return
+	}
+
+	client.SendMessage(smsForm.Parameters, smsForm.Receivers...)
```

Comment:
Check whether receiver is a valid phone number (now we only check for CN, but consider the interface for different countries).